python/triton/compiler.py

```diff
@@ -1306,6 +1275,49 @@ def make_fn_cache_key(fn_hash, signature, configs, constants, num_warps, num_sta
     return key
 
 
+def _compile(fn, signature: str, device: int = -1, constants=dict(), specialization=instance_descriptor(), num_warps: int = 4, num_stages: int = 3, extern_libs=None, output: str = "ttgir") -> Tuple[str, int, str]:
+    valid_outputs = ("ttir", "ttgir", "ptx", "cubin")
+    assert output in valid_outputs, "output should be one of [%s], but get \"%s\"" % (','.join(valid_outputs), output)
+    # triton-ir
+    module = make_triton_ir(fn, signature, specialization, constants)
+    ttir = module.str()
+    if output == "ttir":
+        return ttir
+    # tritongpu-ir
+    module = make_tritongpu_ir(module, num_warps, num_stages)
+    ttgir = module.str()
+    if output == "ttgir":
+        return ttgir
+    # llvm
+    llvm_ir = make_llvm_ir(module, extern_libs)
+    shem_size = _triton.get_shared_memory_size(module)
+    # ptx
+    ptx, kernel_name = make_ptx(llvm_ir, device)
+    if output == "ptx":
+        return ptx, shem_size, kernel_name
+    # cubin
+    cubin = make_cubin(ptx, device)
+    if output == "cubin":
+        return cubin, ptx, ttgir, ttir, shem_size, kernel_name
+    assert False
+
+def read_or_execute(cache_manager, force_compile, file_name, metadata,
+                    run_if_found: Callable[[str], bytes] = None,
+                    run_if_not_found: Callable = None):
+    if not force_compile and cache_manager.has_file(file_name):
+      module = run_if_found(cache_manager._make_path(file_name))
+      data = module if isinstance(module, bytes) else str(module).encode("utf-8")
+      md5 = hashlib.md5(data).hexdigest()
+      suffix = file_name.split(".")[1]
+      has_changed = metadata and md5 != metadata["md5"][suffix]
+      return module, md5, has_changed, True
+    module = run_if_not_found()
+    data = module if isinstance(module, bytes) else str(module).encode("utf-8")
+    md5 = hashlib.md5(data).hexdigest()
+    cache_manager.put(data, file_name, True)
+    return module, md5, True, False
+
+
 def compile(fn, signature: str, device: int = -1, constants=dict(), num_warps: int = 4, num_stages: int = 3, extern_libs=None, configs=None):
```

Comment:
So if a users has a custom IR, he should call the `compile` function and specify the signature manually.

Reply:
if a user has a custom IR, they should really be using `triton.tools.aot` like our unit tests

Reply:
Got you, that makes sense